### PR TITLE
OKTA-308568: Fixed date formatting in query string builder

### PR DESCRIPTION
**Pull request by arvindkrishnakumar-okta**
**Background:** 

See https://github.com/okta/okta-commons-java/issues/35.

### CODE REVIEW

File: http/http-api/src/main/java/com/okta/commons/http/RequestUtils.java [hunk 1]
```diff
@@ -19,13 +19,19 @@
...
+import java.time.format.DateTimeFormatter;
 import java.util.Locale;
 
 /**
  * @since 0.5.0
  */
 public final class RequestUtils {
 
+    private static final DateTimeFormatter DATE_TIME_FORMATTER =
+        DateTimeFormatter.ofPattern("EEE MMM d HH:mm:ss zzz yyyy", Locale.US);
```
Comment (bdemers) [inline http/http-api/src/main/java/com/okta/commons/http/RequestUtils.java:33]: Don't hardcode a locale, you probably want something like:
```suggestion
        DateTimeFormatter.ISO_DATE_TIME.withZone(ZoneOffset.UTC);
```

Comment (arvindkrishnakumar-okta) [inline http/http-api/src/main/java/com/okta/commons/http/RequestUtils.java:33]: The default `java.util.Date` format.

File: http/http-api/src/main/java/com/okta/commons/http/QueryString.java [hunk 2]
```diff
@@ -35,18 +43,29 @@ public QueryString(){}
...
         if (!Collections.isEmpty(source)) {
-            for(Map.Entry<String,?> entry : source.entrySet()) {
+            for (Map.Entry<String,?> entry : source.entrySet()) {
                 String key = entry.getKey();
                 Object value = entry.getValue();
-                String sValue = value != null ? String.valueOf(value) : null;
+                String sValue = null;
+                if (value != null) {
+                    if (value instanceof Date) {
+                        sValue = RequestUtils.getFormattedDate(String.valueOf(value));
```
Comment (bdemers) [inline http/http-api/src/main/java/com/okta/commons/http/QueryString.java:47]: Why is there a `String.valueOf(value)` here? Keep the value a date until it is parsed
```suggestion
                        sValue = RequestUtils.getFormattedDate((Date) value);
```

File: http/http-api/src/main/java/com/okta/commons/http/QueryString.java [hunk 3]
```diff
@@ -35,18 +43,29 @@ public QueryString(){}
...
+                }
                 put(key, sValue);
             }
         }
     }
 
     public String put(String key, Object value) {
         if (value != null) {
+            if (value instanceof Date) {
+                return RequestUtils.getFormattedDate(String.valueOf(value));
```
Comment (bdemers) [inline http/http-api/src/main/java/com/okta/commons/http/QueryString.java:60]: same as above, though you may want to move all of the `toString()` logic into a new method

File: http/http-api/src/test/groovy/com/okta/commons/http/QueryStringTest.groovy [hunk 4]
```diff
@@ -90,4 +90,13 @@ class QueryStringTest {
...
         assertEquals qs.toString(), "key1=value1&key2=value2"
     }
+
+    @Test
+    void testWithDateObjectAsValue() {
+        Date sinceDate = new Date("Thu Nov 30 13:15:16 PST 2017")
+        def query = [since:sinceDate]
+        def qs = new QueryString(query)
+        assertEquals qs.toString(), "since=2017-11-30T21%3A15%3A16Z"
+        assertEquals qs.toString(), "since=" + URLEncoder.encode("2017-11-30T21:15:16Z", "UTF-8")
```
Comment (bdemers) [inline http/http-api/src/test/groovy/com/okta/commons/http/QueryStringTest.groovy:100]: nit: hardcode the encoding here similar to the previous line

File: http/http-api/src/main/java/com/okta/commons/http/RequestUtils.java (hunk 1 above)
Comment (arvindkrishnakumar-okta) [reply http/http-api/src/main/java/com/okta/commons/http/RequestUtils.java]: It doesn't work with this change.  Will need to check why.

File: http/http-api/src/test/groovy/com/okta/commons/http/QueryStringTest.groovy (hunk 4 above)
Comment (arvindkrishnakumar-okta) [reply http/http-api/src/test/groovy/com/okta/commons/http/QueryStringTest.groovy:100]: Not sure if I got what you meant. Line #100 was coded to indicate clearly that line #99 is an encoded variant of the content in line #100. If its confusing, I can remove #100 in its totality?

File: http/http-api/src/main/java/com/okta/commons/http/QueryString.java (hunk 3 above)
Comment (arvindkrishnakumar-okta) [reply http/http-api/src/main/java/com/okta/commons/http/QueryString.java:60]: done

File: http/http-api/src/test/groovy/com/okta/commons/http/QueryStringTest.groovy (hunk 4 above)
Comment (bdemers) [reply http/http-api/src/test/groovy/com/okta/commons/http/QueryStringTest.groovy:100]: oh, i see, yes, plase do

File: http/http-api/src/main/java/com/okta/commons/http/RequestUtils.java (hunk 1 above)
Comment (arvindkrishnakumar-okta) [reply http/http-api/src/main/java/com/okta/commons/http/RequestUtils.java]: Updated, works now.

File: http/http-api/src/test/groovy/com/okta/commons/http/QueryStringTest.groovy (hunk 4 above)
Comment (arvindkrishnakumar-okta) [reply http/http-api/src/test/groovy/com/okta/commons/http/QueryStringTest.groovy]: done

File: http/http-api/src/main/java/com/okta/commons/http/QueryString.java [hunk 5]
```diff
@@ -35,18 +36,29 @@ public QueryString(){}
...
                 Object value = entry.getValue();
-                String sValue = value != null ? String.valueOf(value) : null;
+                String sValue = null;
+                if (value != null) {
+                    if (value instanceof Date) {
+                        sValue = RequestUtils.getFormattedDate((Date) value);
+                    } else {
+                        sValue = String.valueOf(value);
+                    }
+                }
```
Comment (bdemers) [inline http/http-api/src/main/java/com/okta/commons/http/QueryString.java:51]: nit: move the duplicate to string logic to a single method and reuse it in `put` below

Comment (arvindkrishnakumar-okta) [reply http/http-api/src/main/java/com/okta/commons/http/QueryString.java]: Done

**Comment by arvindkrishnakumar-okta** [comment]
> Looks good! one more small change requested

Update pushed.

File: http/http-api/src/main/java/com/okta/commons/http/QueryString.java [hunk 6]
```diff
@@ -35,21 +36,30 @@ public QueryString(){}
...
     }
 
     public String put(String key, Object value) {
+        return super.put(key, getFormattedValue(value));
+    }
+
+    private String getFormattedValue(Object value) {
+        String result = null;
+
         if (value != null) {
```
Comment (bdemers) [inline http/http-api/src/main/java/com/okta/commons/http/QueryString.java:54]: Is this null check lost on the `put`? For example (if i'm reading the code correctly), the old behavior would not add a null value, but the new one will?

Comment (arvindkrishnakumar-okta) [reply http/http-api/src/main/java/com/okta/commons/http/QueryString.java:57]: Good catch! updated code to prevent this.

**Pull request merged**